src/docs/sphinx/developerGuide/Contributing/CodeRules.rst

```diff
@@ -0,0 +1,1440 @@
+        GEOS_LOG(GEOS_FMT("processing elem {}/{}", ei, numElems)); // Will crash on CUDA builds (unsupported GEOS_FMT)
+        printf("DEBUG: ei=%d\n", ei); // Compiles, but remove before merge to develop!
+      });
+
+Structured Data Output
+----------------------
+
+For tabular or statistical output, use structured logging facilities instead of plain text logging:
+
+- Use ``TableTextFormatter`` with a ``logLevel`` XML setting for formatted text table output,
+- Use ``TableCSVFormatter`` with a ``writeCSV()`` XML setting for structured data output.
+
+.. dropdown:: Example: Log table output
+   :icon: code
+
+   .. code-block:: c++
+      // log output preparation, layout precomputation can be kept for periodic outputs
+      TableLayout statsLogLayout( "", { "Flux(es)", "Region", "Element Count", massColumn, rateColumn } );
+      m_logLayout = statsLogLayout;
+
+      // filling table data (log format is for current timestep)
+      m_logData.addRow( fluxName,
+                             elementSetName,
+                             GEOS_FMT( "{}", wrappedStats.stats().m_elementCount ),
+                             GEOS_FMT( "{}", wrappedStats.stats().m_producedMass ),
+                             GEOS_FMT( "{}", wrappedStats.stats().m_productionRate ) );
+
+      // statistics CSV output (only if enabled and from rank 0)
+      if( logLevelActive && MpiWrapper::commRank() == 0 )
+      {
+        string const title = GEOS_FMT( "{}, flux statistics for: {}",
+                                      getName(), fluxesStr );
+        string const multilineTitle = wrapTextToMaxLength( title, 80 ); // enabling automatic line wrapping
+        m_logLayout.setTitle( multilineTitle );
+        TableTextFormatter const tableStatFormatter( m_logLayout );
+        GEOS_LOG_RANK( tableStatFormatter.toString( statsData ) );
+      }
+
+.. dropdown:: Example: CSV output
+   :icon: code
+
+   .. code-block:: c++
+
+      // CSV output preparation, layout precomputation can be kept for periodic outputs
+      TableLayout const statsCSVLayout( "", {"Time [s]", "Flux(es)", "Region", "Element Count", massColumn, rateColumn} );
+      m_csvLayout = statsCSVLayout;
+
+      // filling table data (CSV format is all timestep series)
+      m_csvData.addRow( wrappedStats.getStatsPeriodStart(),
+                             fluxName,
+                             elementSetName,
+                             wrappedStats.stats().m_elementCount,
+                             wrappedStats.stats().m_producedMass,
+                             wrappedStats.stats().m_productionRate );
+
+      // statistics CSV output (only if enabled and from rank 0)
+      if( m_writeCSV > 0 && MpiWrapper::commRank() == 0 )
+      {
+        std::ofstream outputFile( m_csvFilename );
+        TableCSVFormatter const tableStatFormatter( m_csvLayout );
+        outputFile << tableStatFormatter.toString( csvData );
```

Comment:
Probably a side not, but it is confusing that std's `operator<<` is used here while unadvised (generally speaking) elsewhere, but the answer is probably in the code not in the doc